nexus/reconfigurator/planning/src/blueprint_editor/sled_editor/underlay_ip_allocator.rs

```diff
@@ -0,0 +1,220 @@
+///
+/// The current implementation takes the max address seen so far and uses the
+/// next one.  This will never reuse old IPs.  That avoids a bunch of
+/// operational issues.  It does mean we will eventually run out of IPs.  But we
+/// do have a big space right now (2^16).
+// This overlaps with the bump allocator that's used in RSS.  That one is not
+// general enough to use here, though this one could potentially be used there.
+#[derive(Debug)]
+pub(crate) struct SledUnderlayIpAllocator {
+    minimum: Ipv6Addr,
+    last: Ipv6Addr,
+    maximum: Ipv6Addr,
+}
+
+impl SledUnderlayIpAllocator {
+    /// Create a new allocator for the given sled subnet that reserves all the
+    /// specified IPs.
+    ///
+    /// Fails if any of the specified IPs are not part of the sled subnet.
+    pub fn new<I>(
+        sled_subnet: Ipv6Subnet<SLED_PREFIX>,
+        in_use_zone_ips: I,
+    ) -> Result<Self, SledUnderlayIpOutOfRange>
+    where
+        I: Iterator<Item = (ZoneKind, Ipv6Addr)>,
+    {
+        let sled_subnet_addr = sled_subnet.net().prefix();
+        let minimum = sled_subnet_addr
+            .saturating_add(u128::from(SLED_RESERVED_ADDRESSES));
+        let maximum = sled_subnet_addr
+            .saturating_add(u128::from(CP_SERVICES_RESERVED_ADDRESSES));
+        assert!(maximum > minimum);
+
+        // We shouldn't need to explicitly reserve the sled's global
+        // zone and switch addresses because they should be out of our
+        // range, but we do so just to be sure.
+        let sled_gz_addr = *get_sled_address(sled_subnet).ip();
+        assert!(sled_subnet.net().contains(sled_gz_addr));
+        assert!(minimum > sled_gz_addr);
+        let switch_zone_addr = get_switch_zone_address(sled_subnet);
+        assert!(sled_subnet.net().contains(switch_zone_addr));
+        assert!(minimum > switch_zone_addr);
+        assert!(sled_subnet.net().contains(minimum));
+        assert!(sled_subnet.net().contains(maximum));
+
+        let mut slf = Self { minimum, last: minimum, maximum };
+        for (zone_kind, ip) in in_use_zone_ips {
+            slf.mark_as_allocated(zone_kind, ip)?;
+        }
+        assert!(slf.minimum <= slf.last);
+        assert!(slf.last <= slf.maximum);
+
+        Ok(slf)
+    }
+
+    /// Mark an address as used.
+    ///
+    /// Marking an address that has already been handed out by this allocator
+    /// (or could have been handed out by this allocator) is allowed and does
+    /// nothing.
+    ///
+    /// # Errors
+    ///
+    /// Fails if `ip` is outside the subnet of this sled.
+    pub fn mark_as_allocated(
+        &mut self,
+        zone_kind: ZoneKind,
+        ip: Ipv6Addr,
+    ) -> Result<(), SledUnderlayIpOutOfRange> {
+        // We intentionally ignore any internal DNS underlay IPs; they live
+        // outside the sled subnet and are allocated separately.
+        if zone_kind == ZoneKind::InternalDns {
+            return Ok(());
+        }
```

Comment:
I see.  My first reaction to this was: this is an allocator for IPs from the *sled's* subnet.  Callers should know if that's what they have and they should simply avoid calling this for the DNS IP, which would be fairly obviously _not_ from the sled's subnet.

Looking at the call stack though, that isn't the case.  `BlueprintZoneConfig::underlay_ip()` hides this detail.

It kind of makes me wonder if we want a sort of smarter super-allocator where you can gave it an enum like SledUnderlayIp, InternalDnsUnderlayIp, maybe even fold NexusExternalIp and ExternalDnsExternalIp into it, and then you'd specify which thing you wanted when you mark things as used or allocated.  But I haven't thought about this much -- it's pretty speculative -- and it seems way out of scope just to avoid this check!  So I say go with this for now.

Reply:
I think we can check against the reserved rack subnet. That amounts to the same thing as this check, but has better semantics. It also lets us drop `zone_kind`, which is a nice simplification. I'm most of the way through a patch for this, if you think it's a good idea I can push to this branch.

Reply:
> It kind of makes me wonder if we want a sort of smarter super-allocator where you can gave it an enum like SledUnderlayIp, InternalDnsUnderlayIp, maybe even fold NexusExternalIp and ExternalDnsExternalIp into it, and then you'd specify which thing you wanted when you mark things as used or allocated.

This sounds nice to me at first glance, but I'm not sure about the mechanics. Sled underlay IP decisions can be made locally (with just the information about the one sled in question), but the other three need information about all sleds. I'll think about this one (but agreed that'd be a followup PR).

> I think we can check against the reserved rack subnet. That amounts to the same thing as this check, but I think it has better semantics. It also lets us drop `zone_kind`, which is a nice simplification.

I'm less sure about this - if we didn't pass in `zone_kind`, would we allow a mismatch (e.g., a Nexus zone using an IP from the reserved rack subnet)? Maybe that kind of thing should be caught by a different level (e.g., blippy, which I'm hoping to have a draft out for soon), and then it would be reasonable to simplify this as you suggest?